### ISSUE TITLE
[WEIRD_BEHAVIOUR/BUG] possible bspwm "infinite loop" caused by setting a rule and common external rules practice.

### ISSUE THREAD

**Issue by emanuele6** [baskerville/bspwm#1162]
**output of `bspwm -v`**: `0.9.9-28-g9ef79d7`
**content of bspwmrc**: an empty one is fine.

It is common practice to use `eval "$4"` or `eval $4` in `external_rules_command` scripts written in a shell language to parse rule consequences.

Let's assume the content of `external_rules_command` is the following, for these examples:
```sh
#!/bin/sh

notify-send "$4"
eval "$4"
```

I was writing an external rule in my `external_rules_command` script, I noticed `eval "$4"` and I wondered if I could make `bspwm` run commands by setting a rule.

I tried to run `bspc rule -a '*' -o node='$(hellof)'` where `hellof` is this a script (I knew I couldn't use spaces in rules so `notify-send hello` wouldn't have worked):

```sh
#!/bin/sh

notify-send hello
```

Hilariously, it worked!

After a bit, I tried doing it again, but this time I used the `state` key (I was using keys at random, I didn't pick a different one on purpose) instead of `node` and I noticed it didn't work. So, I opened the `bspwm(1)` man-page and tried to do this with all the keys to see which ones worked: only `monitor`, `desktop`, `node` and, surprisingly, `split_dir`.
  
Now, I was wondering what would happen if I were to make it run a command which creates a window and waits for the window to close before exiting like `urxvt` and I were to make the rule non-`one-shot`, i.e.:

```sh
bspc node -a '*' node='$(urxvt)'
```

After running that command, I pressed my keybind to launch a terminal and I started getting a lot of notifications from `external_rules_command` non-stop and no window was getting shown.

--------------------------------------------------

Even though I realise that this exploit to make `bspwm` run commands before managing windows is not that harmful, I don't see the point in keeping it since it is limited to commands without spaces and it's unintended and just weird.

This is pretty easy to fix for `split_dir` (see #1161): just store `split_dir` as a `direction_t *` instead of as a string in rule consequences (in the same way `layer` is stored as a `stack_layer_t *` and `state` is stored as a `client_state_t *`) . This also prevents bspwm from passing an invalid `split_dir` to `external_rules_command`.

`node`, `desktop` and `monitor` could contain "$(hellof)" and still be valid since "$(hellof)" could be a monitor name for `monitor` or it could be either a monitor name or a desktop name for `node` and `desktop`.
I think that the good way to solve this is making `bspwm` query the `NODE_SEL`s, `DESKTOP_SEL`s and `MONITOR_SEL`s before passing them to `external_rules_command` (EDIT: maybe even before storing in the `rule_consequence_t` variable so that `{monitor,desktop,node}_desc` can just be stored as ids instead of as `char [MAXLEN]`s) and just passing the ids instead of the whole selectors (which are probably more useful anyway) or nothing if the query was not successful.

If the value of `desktop` is a `DESKTOP_SEL` which is invalid or that yields no results, an external rule like the following wouldn't work (you probably wouldn't expect it to not work):
```sh
#!/bin/sh

eval "$4"

bspc desktop "${desktop:-focused}" -n "$(date +'%I|%M|%S%p')"
```

Some example rule to test the previous example:
```shell
bspc rule -a '*' -o desktop='any.focused.!focused' # never yields results
bspc rule -a '*' -o desktop='.occupied'            # invalid desktop selector, because not a descriptor
bspc rule -a '*' -o desktop='hello:hello:'         # invalid desktop selector
```

**Comment by emanuele6** [comment]
This issue is also about making sure that the consequences argument passed to `external_rules_command` always has only valid values.

**Comment by baskerville** [comment]
> ```shell
> bspc rule -a '*' -o desktop='any.focused.!focused' # never yields results
> bspc rule -a '*' -o desktop='.occupied'            # invalid desktop selector, because not a descriptor
> bspc rule -a '*' -o desktop='hello:hello:'         # invalid desktop selector
> ```

You might want to create a separate issue for this: those mistakes can be caught during the parsing stage.

**Comment by emanuele6** [comment]
Fixed by 7313b01 (`split_dir`) and 5d93323 (`monitor`, `desktop`, `node`).
